File: app/src/main/java/com/example/yangweichao/lifecycles/MyLocationListener.java
package com.example.yangweichao.lifecycles;

import android.arch.lifecycle.Lifecycle;
import android.arch.lifecycle.LifecycleObserver;
import android.arch.lifecycle.OnLifecycleEvent;
import android.content.Context;

import static android.arch.lifecycle.Lifecycle.State.STARTED;

class MyLocationListener implements LifecycleObserver {
    private boolean enabled = false;
    public MyLocationListener(Context context, Lifecycle lifecycle, DemoCallBack callback) {

    }

    @OnLifecycleEvent(Lifecycle.Event.ON_START)
    void start() {
        if (enabled) {
           // connect
        }
    }

    public void enable() {
        enabled = true;
//        if (lifecycle.getCurrentState().isAtLeast(STARTED)) {
//            // connect if not connected
//        }
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_STOP)
    void stop() {
        // disconnect if connected
    }
}

File: app/src/main/java/com/example/yangweichao/lifecycles/MylocationListener.kt
package com.example.yangweichao.lifecycles

import android.content.Context
import javax.security.auth.callback.Callback

/**
 * Created by yangweichao on 2018/4/8.
 */
open  class  MylocationListener constructor (mainActivity: MainActivity, nothing: DemoCallBack) {



    fun start() {
        // connect to system location service
    }

    fun stop() {
        // disconnect from system location service
    }
}